modules/io/src/writers/xviz-protobuf-writer.js

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/* eslint-disable */
+import {XVIZBaseWriter} from './xviz-base-writer';
+import {XVIZ_PROTOBUF_MAGIC, XVIZ_PROTOBUF_MESSAGE} from '../common/protobuf-support';
+
+// 0-frame is an index file for timestamp metadata
+// 1-frame is the metadata file for the log
+// 2-frame is where the actual XVIZ updates begin
+const messageName = index => `${index + 2}-frame`;
+
+export class XVIZProtobufWriter extends XVIZBaseWriter {
+  constructor(sink, options = {}) {
+    super(sink);
+
+    const {envelope = true} = options;
+    this.messageTimings = {
+      messages: new Map()
+    };
+    this.wroteMessageIndex = null;
+    this.options = {envelope};
+  }
+
+  // xvizMetadata is the object returned
+  // from a Builder.
+  writeMetadata(xvizMetadata) {
+    this._checkValid();
+    this._saveTimestamp(xvizMetadata);
+
+    const pbJSON = xvizConvertProtobuf(xvizMetadata);
+    let pbType = XVIZ_PROTOBUF_MESSAGE.Metadata;
+    let pbMsg = pbType.fromObject(pbJSON);
+
+    if (this.options.envelope) {
+      const value = pbType.encode(pbMsg).finish();
+      pbType = XVIZ_PROTOBUF_MESSAGE.Envelope;
+      pbMsg = pbType.fromObject({
+        type: 'xviz/metadata',
+        data: {type_url: 'xviz.v2.Metadata', value}
+      });
+    }
+
+    const pbBuffer = pbType.encode(pbMsg).finish();
+    const buffer = new Uint8Array(pbBuffer.byteLength + 4);
+    buffer.set(XVIZ_PROTOBUF_MAGIC, 0);
+    buffer.set(pbBuffer, 4);
+    this.writeToSink('1-frame.pbe', buffer);
+  }
+
+  writeMessage(messageIndex, xvizMessage) {
+    this._checkValid();
+    this._saveTimestamp(xvizMessage, messageIndex);
+
+    const pbJSON = xvizConvertProtobuf(xvizMessage);
+    let pbType = XVIZ_PROTOBUF_MESSAGE.StateUpdate;
+    let pbMsg = pbType.fromObject(pbJSON);
+
+    if (this.options.envelope) {
```

Comment:
This enveloping looks like it should be in a helper function since it's used below too and a bit complex.

Reply:
Done